codec-dns/src/main/java/io/netty/handler/codec/dns/AbstractDnsMessage.java

```diff
@@ -0,0 +1,464 @@
+     */
+    protected AbstractDnsMessage(int id) {
+        this(id, DnsOpCode.QUERY);
+    }
+
+    /**
+     * Creates a new instance with the specified {@code id} and {@code opCode}.
+     */
+    protected AbstractDnsMessage(int id, DnsOpCode opCode) {
+        setId(id);
+        setOpCode(opCode);
+    }
+
+    @Override
+    public int id() {
+        return id & 0xFFFF;
+    }
+
+    @Override
+    public DnsMessage setId(int id) {
+        this.id = (short) id;
+        return this;
+    }
+
+    @Override
+    public DnsOpCode opCode() {
+        return opCode;
+    }
+
+    @Override
+    public DnsMessage setOpCode(DnsOpCode opCode) {
+        this.opCode = checkNotNull(opCode, "opCode");
+        return this;
+    }
+
+    @Override
+    public boolean isRecursionDesired() {
+        return recursionDesired;
+    }
+
+    @Override
+    public DnsMessage setRecursionDesired(boolean recursionDesired) {
+        this.recursionDesired = recursionDesired;
+        return this;
+    }
+
+    @Override
+    public int z() {
+        return z;
+    }
+
+    @Override
+    public DnsMessage setZ(int z) {
+        this.z = (byte) (z & 7);
+        return this;
+    }
+
+    @Override
+    public int count(DnsSection section) {
+        return count(sectionOrdinal(section));
+    }
+
+    private int count(int section) {
+        final Object records = sectionAt(section);
+        if (records == null) {
+            return 0;
+        }
+        if (records instanceof DnsRecord) {
+            return 1;
+        }
+
+        @SuppressWarnings("unchecked")
+        final List<DnsRecord> recordList = (List<DnsRecord>) records;
+        return recordList.size();
+    }
+
+    @Override
+    public int count() {
+        int count = 0;
+        for (int i = 0; i < SECTION_COUNT; i ++) {
+            count += count(i);
+        }
+        return count;
+    }
+
+    @Override
+    public <T extends DnsRecord> T recordAt(DnsSection section) {
+        return recordAt(sectionOrdinal(section));
+    }
+
+    private <T extends DnsRecord> T recordAt(int section) {
+        final Object records = sectionAt(section);
+        if (records == null) {
+            return null;
+        }
+
+        if (records instanceof DnsRecord) {
+            return castRecord(records);
+        }
+
+        @SuppressWarnings("unchecked")
+        final List<DnsRecord> recordList = (List<DnsRecord>) records;
+        if (recordList.isEmpty()) {
+            return null;
+        }
+
+        return castRecord(recordList.get(0));
+    }
+
+    @Override
+    public <T extends DnsRecord> T recordAt(DnsSection section, int index) {
+        return recordAt(sectionOrdinal(section), index);
+    }
+
+    private <T extends DnsRecord> T recordAt(int section, int index) {
+        final Object records = sectionAt(section);
+        if (records == null) {
+            throw new IndexOutOfBoundsException(String.valueOf(index));
```

Comment:
What about a more meaningful message for these exceptions. Like include what range is valid. WDYT?